### ISSUE TITLE
Add cache date-of-publication

### ISSUE THREAD

**Issue by rapotek** [opencaching/opencaching-pl#1386]
Enhancement proposal: when the cache for the first time is made available (ready) the date and time of the moment should be stored in the caches table, in the separate column named f.ex. `date_published`. The column should be updated only once per cache, internally and accessed from outside in read only mode. The column should be default null and can be made indexed too, for future usability in search and sorting.

The column modifications could be made even using the SQL itself only, f.ex:
in INSERT values - `(CASE :status WHEN 1 THEN NOW() ELSE NULL END)`,
in UPDATE - `SET date_published = CASE WHEN :status = 1 AND date_published IS NULL THEN NOW() ELSE date_published END`

This proposal could be taken into consideration f.ex. in case of `newcache.php` and `editcache.php` code mess refactoring, but other code elements changing the cache status, if there are any, should be as well updated then.

**Comment by kojoty** [comment]
I edit a subject a little and I agree with this postulate.

**Comment by kojoty** [comment]
in fact this is duplicate of #426

**Comment by rapotek** [comment]
I do not know much about administrative functions in the service, so I have a question: is it possible for cache to have status=6 (blocked) before it has been visible? I mean is it possible to achieve in the services change of status from 4 or 5 to 6?
It may be not so obvious but the question is related to this issue.

**Comment by kojoty** [comment]
@rapotek, yeah, the state-diagram for cache is quite complicated - I've tried to prepare the docs for it - this is what I manage to collect:
https://docs.google.com/spreadsheets/d/XXX1BpkpmusS98HQKmyJJpJhCQtKHOltdY1b7QE6f8poW8E/edit#gid=0 (please remove XX form the link)

**Comment by rapotek** [comment]
@kojoty thanks, I will analyze it.

**Comment by deg-pl** [comment]
Kiedyś Sebastian chciał opublikować kesza, gdzie na finale on (Sebastian) miał wisieć... na drzewie. Takie "ostatnie pożegnanie".
COG zablokował kesza przed publikacją.
Komentarz dodaję w celu wytłumaczenia dlaczego czasami bywa taka potrzeba... Sebastian żyje do dzisiaj ;)

**Comment by rapotek** [comment]
> Sebastian żyje do dzisiaj ;)
Zauważyłem... ;)

Ogólnie: mam pewien bardzo prosty pomysł, ale muszę go jeszcze przemyśleć. Diagram przejść stanów mi się w tym przyda.

**Comment by rapotek** [comment]
I do not like triggers but this seems to be the least complicated way to introduce date-of-publication avoiding php code modifications.

See [this commit](https://github.com/rapotek/opencaching-pl/commit/ec0e158f75461f4d60c0533f7b9f1a5513b20ec7). It alters `caches` table adding `date_published` column, updates existing caches filling up `date_published` wherever it can be estimated and modifies existing triggers to provide the valid data in correct status change. The updates and triggers has been tested on my developement database with a few test caches created.

The update existing caches procedure is described in new sqlAlters file, but will try to explain it shortly.
One explanation first: the `date_hidden` value can be changed by owner in any time to, more or less, any value, so it alone cannot be the determinant of publication date.

The existing caches being in "visible" states 1, 2 or 3 are treated as published, so it is desirable to fill the `date_published` value there. If the cache has any non-admin log (even deleted), the least recent log `date_created` is compared with the cache `date_hidden`. Next the earlier datetime of these two is compared with the cache `date_created` where the more recent datetime wins. In result `date_published` should be always between cache `date_created` and the first non_admin log `date_created` regardless of `date_hidden` settings.

The existing caches being in "blocked" state 6 could be earlier in any state, the "visible" ones too, but it cannot be determined. So `date_published` value is set only for the cache having any non-admin log related. In this case the similar algorithm as for "visible" caches applies.

The updates described above are not being perfect but I'm afraid there is not enough information stored to estimate better datetime values. I considered to check `last_modified` value too, but it can change in any moment, before or after publication, so I would not trust it in the case.

The modified triggers sets `date_published` to `now()` before insert or update when the `status` value set is 1 or 2 ( the latter is possibly not necessary for insert) and for update `date_published` is set only if it was not set earlier. In this case cache created and archived before being 'visible' will have a null value for `date_published`.

**Comment by andrixnet** [comment]
The first 2 sheets are understandable, but the flowchart (image) is PL only.

**Comment by kojoty** [comment]
Sorry @andrixnet, the third sheet is a work of our cog member Jasiek - this is result of his research. This is not a final version - I' m going to work at this in future - but there are so much other issues :)

**Comment by DorotkaMakotka** [comment]
> Kiedyś Sebastian chciał opublikować kesza, 
Serio? 
Dzieci drogie, czy naprawdę te oto otwarte forum jest dobrym miejscem na tego typu historie?
Pomijając etyczną stronę propagowania sensacji, mało to profesjonalne.
Ploteczki i inne obyczajowe skandale lepiej chyba roztrząsać sobie gdzieś prywatnie.
:)
Pozdrawiam,
Ciocia makotka

**Comment by rapotek** [comment]
I suppose this issue can be closed but I think it needs confirmation that the date of publication works correctly on each node.

**Comment by kojoty** [comment]
@andrixnet, @harrieklomp, 
please confirm that publication date in caches created at your nodes before april 2018 are reasonable - we have some problems with it at OCPL  - most of caches have wrong publ.-date - now we fixed it in our DB.

Please reopen if there is anything more to do.
